pkg/fqdn/dnsproxy/proxy_test.go

```diff
@@ -163,131 +206,325 @@ func (s *DNSProxyTestSuite) TearDownSuite(c *C) {
+	err = s.proxy.UpdateAllowed(epID_1, dstPort, l7map)
+	c.Assert(err, Equals, nil, Commentf("Could not update with rules"))
+	allowed, err := s.proxy.CheckAllowed(epID_1, dstPort, dstID_1, query)
+	c.Assert(err, Equals, nil, Commentf("Error when checking allowed"))
+	c.Assert(allowed, Equals, true, Commentf("request was rejected when it should be allowed"))
+
+	// Delete once, it should reject
+	err = s.proxy.UpdateAllowed(epID_1, dstPort, nil)
+	c.Assert(err, Equals, nil, Commentf("Could not update with rules"))
+	allowed, err = s.proxy.CheckAllowed(epID_1, dstPort, dstID_1, query)
+	c.Assert(err, Equals, nil, Commentf("Error when checking allowed"))
+	c.Assert(allowed, Equals, false, Commentf("request was allowed when it should be rejected"))
+
+	// Delete once, it should reject and not crash
+	err = s.proxy.UpdateAllowed(epID_1, dstPort, nil)
+	c.Assert(err, Equals, nil, Commentf("Could not update with rules"))
+	allowed, err = s.proxy.CheckAllowed(epID_1, dstPort, dstID_1, query)
+	c.Assert(err, Equals, nil, Commentf("Error when checking allowed"))
+	c.Assert(allowed, Equals, false, Commentf("request was allowed when it should be rejected"))
 }
 
-func (s *DNSProxyTestSuite) TestSetRejectReplyNoValidData(c *C) {
-	s.proxy.SetRejectReply("banana")
-	request := new(dns.Msg)
-	request.SetQuestion("notcilium.io.", dns.TypeA)
-	response, _, err := s.dnsTCPClient.Exchange(request, s.proxy.TCPServer.Listener.Addr().String())
-	c.Assert(err, IsNil, Commentf("DNS request from test client returned error when it should be rejected"))
-	c.Assert(response.Rcode, Not(Equals), 100, Commentf("DNS request from test client has an invalid response code"))
+func (s *DNSProxyTestSuite) TestFullPathDependence(c *C) {
+	// Test that we consider each of endpoint ID, destination SecID (via the
+	// selector in L7DataMap), destination port (set in the redirect itself) and
+	// the DNS name.
+	// The rules approximate:
+	//	+------+--------+---------+----------------+
+	//	| From |   To   | DstPort |    DNSNames    |
+	//	+======+========+=========+================+
+	//	| All  | DstID1 |      53 | *.ubuntu.com   |
+	//	| All  | DstID1 |      53 | aws.amazon.com |
+	//	| All  | DstID2 |      53 | cilium.io      |
+	//	| All  | DstID1 |      54 | example.com    |
+	//	+------+--------+---------+----------------+
+	//
+	// Cases:
+	// +------+-------+--------+------+-------------+----------+-------------------------------------+
+	// | Case | From  |   To   | Port |    Query    | Outcome  |               Reason                |
+	// +------+-------+--------+------+-------------+----------+-------------------------------------+
+	// |    1 | EPID1 | DstID1 |   53 | cilium.io   | Allowed  |                                     |
+	// |    2 | EPID1 | DstID1 |   54 | cilium.io   | Rejected | Port 54 only allows example.com     |
+	// |    3 | EPID1 | DstID2 |   53 | cilium.io   | Rejected | DstID2 is not allowed in any policy |
```

Comment:
I don't understand, the table on line 437 states that DstID2 is in fact allowed on port 53 to `cilium.io`, but this table says that such traffic is rejected.

Reply:
Yup, this was a mistake. I'll fixup Dst2 per your other comment but this test should be for aws.amazon.com. It was intended to check for aliasing between L3 selectors (i.e. aws.amazon.com is only allowed for Dst1). The other tests are also a little off, since cilium.io is only allowed with Dst2... but Case 1 thinks it's allowed :/

Reply:
(the same for EPID2. The policies should say From: EPID1 in all cases)